pkg/node/manager.go

```diff
@@ -64,12 +67,195 @@ func deleteNodeCIDR(ip *net.IPNet) {
+		if r.LinkIndex == route.LinkIndex && r.Scope == route.Scope &&
+			aMaskLen == bMaskLen && aMaskBits == bMaskBits &&
+			r.Dst.IP.Equal(route.Dst.IP) && r.Gw.Equal(route.Gw) {
+			return &r
+		}
+	}
+
+	return nil
+}
+
+func replaceNexthopRoute(link netlink.Link, routerIP, routerNet *net.IPNet) error {
+	// Add the Cilium router IP as address to the "cilium_host" if not
+	// already assigned
+	if routerIP != nil && findAddress(link, routerIP.IP) == nil {
+		addr := &netlink.Addr{IPNet: routerIP}
+		if err := netlink.AddrAdd(link, addr); err != nil {
+			return fmt.Errorf("unable to add nexthop address \"%s\": %q", routerIP, err)
+		}
+
+		log.WithFields(log.Fields{
+			logfields.IPAddr:    routerIP,
+			logfields.Interface: link.Attrs().Name,
+		}).Info("Added Cilium router IP address")
+	}
+
+	// This is the L2 route which makes the Cilium router IP available behind
+	// the "cilium_host" interface. All other routes will use this router IP
+	// as nexthop.
+	route := &netlink.Route{
+		LinkIndex: link.Attrs().Index,
+		Dst:       routerNet,
+		Scope:     netlink.SCOPE_LINK,
+	}
+	scopedLog := log.WithField(fieldRoute, route)
+
+	if findRoute(link, route) == nil {
+		if err := netlink.RouteDel(&netlink.Route{
+			LinkIndex: link.Attrs().Index,
+			Dst:       routerNet,
+		}); err != nil {
+			scopedLog.WithError(err).Debug("No old L2 nexthop route exists")
+		}
+
+		if err := netlink.RouteAdd(route); err != nil {
+			scopedLog.WithError(err).Error("Unable to add L2 nexthop route")
+			return fmt.Errorf("unable to add L2 nexthop route")
+		}
+
+		scopedLog.Info("Added L2 nexthop route")
+	}
+
+	return nil
+}
+
+func replaceNodeRoute(ip *net.IPNet) {
+	link, err := netlink.LinkByName("cilium_host")
+	if err != nil {
+		log.WithError(err).Error("Unable to lookup interface cilium_host")
+		return
+	}
+
+	var routerIP, routerNet *net.IPNet
+	var via, local net.IP
+	if ip.IP.To4() != nil {
+		via = net.IPv4(169, 254, 254, 1)
+		routerIP = &net.IPNet{IP: via, Mask: net.CIDRMask(32, 32)}
+		routerNet = &net.IPNet{IP: net.IPv4(169, 254, 254, 0), Mask: net.CIDRMask(24, 32)}
+		local = GetInternalIPv4()
+	} else {
+		via = GetIPv6Router()
+		// IPv6 does not require the router IP to be added as address, leaving
+		routerIP = nil
+		routerNet = &net.IPNet{IP: via, Mask: net.CIDRMask(128, 128)}
+		local = GetIPv6()
+	}
+
+	if err := replaceNexthopRoute(link, routerIP, routerNet); err != nil {
+		log.WithError(err).Error("Unable to add nexthop route")
+	}
+
+	route := netlink.Route{LinkIndex: link.Attrs().Index, Dst: ip, Gw: via, Src: local}
+	scopedLog := log.WithField(fieldRoute, route)
+
+	if err := netlink.RouteDel(&route); err != nil {
+		scopedLog.WithError(err).Debug("Old node route does not exist")
+	}
+
+	if err := netlink.RouteAdd(&route); err != nil {
+		scopedLog.WithError(err).Error("Unable to add node route")
+	} else {
+		scopedLog.Info("Installed node route")
+	}
+}
+
+func installHostRoutes() {
+	if !ciliumHostInitialized {
+		return
+	}
```

Comment:
Since we went to all the effort to guard this (in InstallHostRoutes and here in installHostRoutes) we should log something here, or return an error that can be logged at the callers.

Reply:
The case where we might hit this is if the Kubernetes watcher is installed before we put the `cilium_host` device in place in which case we will start collecting node information without being able to install the routes. They will be installed when we add the device. I'll add a debug statement to indicate that installation was deferred.